math/src/field/fields/p448_goldilocks_prime_field.rs

```diff
@@ -0,0 +1,304 @@
+            bbb[i] = bb[i] + b[i + 4];
+        }
+
+        let widemul = |a: u64, b: u64| -> u128 { (a as u128) * (b as u128) };
+
+        for i in 0..4 {
+            accum2 = 0;
+
+            for j in 0..=i {
+                accum2 += widemul(a[j], b[i - j]);
+                accum1 += widemul(aa[j], bb[i - j]);
+                accum0 += widemul(a[j + 4], b[i - j + 4]);
+            }
+            for j in (i + 1)..4 {
+                accum2 += widemul(a[j], b[8 - (j - i)]);
+                accum1 += widemul(aa[j], bbb[4 - (j - i)]);
+                accum0 += widemul(a[j + 4], bb[4 - (j - i)]);
+            }
+
+            accum1 -= accum2;
+            accum0 += accum2;
+
+            c[i] = (accum0 as u64) & mask;
+            c[i + 4] = (accum1 as u64) & mask;
+
+            accum0 >>= 56;
+            accum1 >>= 56;
+        }
+
+        accum0 += accum1;
+        accum0 += c[4] as u128;
+        accum1 += c[0] as u128;
+        c[4] = (accum0 as u64) & mask;
+        c[0] = (accum1 as u64) & mask;
+
+        accum0 >>= 56;
+        accum1 >>= 56;
+
+        c[5] += accum0 as u64;
+        c[1] += accum1 as u64;
+
+        U56x8 { limbs: c }
+    }
+
+    fn sub(a: &U56x8, b: &U56x8) -> U56x8 {
+        let co1 = ((1u64 << 56) - 1) * 2;
+        let co2 = co1 - 2;
+
+        let mut limbs = [0u64; 8];
+        for (i, limb) in limbs.iter_mut().enumerate() {
+            *limb =
+                a.limbs[i]
+                    .wrapping_sub(b.limbs[i])
+                    .wrapping_add(if i == 4 { co2 } else { co1 });
+        }
+
+        let mut res = U56x8 { limbs };
+        Self::weak_reduce(&mut res);
+        res
+    }
+
+    fn neg(a: &U56x8) -> U56x8 {
+        let zero = Self::zero();
+        Self::sub(&zero, a)
+    }
+
+    fn inv(a: &U56x8) -> U56x8 {
+        Self::pow(a, P448_GOLDILOCKS_PRIME_FIELD_ORDER - U448::from_u64(2))
+    }
+
+    fn div(a: &U56x8, b: &U56x8) -> U56x8 {
+        let b_inv = Self::inv(b);
+        Self::mul(a, &b_inv)
+    }
+
+    fn eq(a: &U56x8, b: &U56x8) -> bool {
+        a.limbs == b.limbs
+    }
+
+    fn zero() -> U56x8 {
+        U56x8 { limbs: [0u64; 8] }
+    }
+
+    fn one() -> U56x8 {
+        let mut limbs = [0u64; 8];
+        limbs[0] = 1;
+        U56x8 { limbs }
+    }
+
+    fn from_u64(x: u64) -> U56x8 {
+        let mut limbs = [0u64; 8];
+        limbs[0] = x & ((1u64 << 56) - 1);
+        limbs[1] = x >> 56;
+        U56x8 { limbs }
+    }
+
+    fn from_base_type(x: U56x8) -> U56x8 {
+        let mut x = x;
+        Self::weak_reduce(&mut x);
+        x
+    }
+}
+
+impl P448GoldilocksPrimeField {
+    fn weak_reduce(a: &mut U56x8) {
+        let a = &mut a.limbs;
+
+        let mask = (1u64 << 56) - 1;
+        let tmp = a[7] >> 56;
+        a[4] += tmp;
+
+        for i in (1..8).rev() {
+            a[i] = (a[i] & mask) + (a[i - 1] >> 56);
+        }
+
+        a[0] = (a[0] & mask) + tmp;
+    }
+}
+
+impl U56x8 {
+    pub const fn from(value: &str) -> Self {
```

Comment:
```suggestion
    pub const fn from_hex_str(value: &str) -> Self {
```

Reply:
This is the same as the rest of our code, we should fix it everywhere

Reply:
Then leave it for later :)